combinetf2/physicsmodels.py

```diff
@@ -0,0 +1,399 @@
+
+        if len(hist_axes) != len(axes_names):
+            raise ValueError(
+                f"Hist axes {[h.name for h in hist_axes]} != {axes_names} not found"
+            )
+
+        exp_axes = channel_axes.copy()
+
+        # for per process histograms
+        exp_axes.append(indata.axis_procs)
+        extra_axes = [indata.axis_procs]
+
+        self.exp_shape = tuple([len(a) for a in exp_axes])
+
+        channel_axes_names = [axis.name for axis in channel_axes]
+        extra_axes_names = [axis.name for axis in extra_axes]
+
+        axis_idxs = [channel_axes_names.index(axis) for axis in axes_names]
+
+        self.proj_idxs = [i for i in range(len(channel_axes)) if i not in axis_idxs]
+
+        post_proj_axes_names = [
+            axis for axis in channel_axes_names if axis in axes_names
+        ] + extra_axes_names
+
+        self.transpose_idxs = [
+            post_proj_axes_names.index(axis) for axis in axes_names
+        ] + [post_proj_axes_names.index(axis) for axis in extra_axes_names]
+
+        self.has_data = not info["masked"]
+
+        self.channel_info = {
+            channel: {
+                "axes": hist_axes,
+                "start": None,
+                "stop": None,
+            }
+        }
+
+        self.identifiers = [
+            self.name,
+            "_".join(axes_names) if len(axes_names) else "yield",
+        ]
+
+    def make_fun(self, fun_flat, inclusive=True):
+
+        if inclusive:
+            out_shape = self.exp_shape[:-1]
+            transpose_idxs = self.transpose_idxs[:-1]
+        else:
+            out_shape = self.exp_shape
+            transpose_idxs = self.transpose_idxs
+
+        def fun():
+            exp = fun_flat()[self.start : self.stop]
+            if self.normalize:
+                norm = tf.reduce_sum(exp)
+                exp /= norm
+            exp = tf.reshape(exp, out_shape)
+            exp = tf.reduce_sum(exp, axis=self.proj_idxs)
+            exp = tf.transpose(exp, perm=transpose_idxs)
+
+            return exp
+
+        return fun
+
+    def get_data(self, data, cov=None):
+        val = self.make_fun(lambda: data)()
+
+        shape = self.exp_shape[:-1]
+        perm_idxs = self.transpose_idxs[:-1]
+
+        if cov is not None:
+            cov = cov[self.start : self.stop, self.start : self.stop]
+
+            cov = tf.reshape(cov, [*shape, *shape])
+            for idx in sorted(self.proj_idxs, reverse=True):
+                cov = tf.reduce_sum(cov, axis=idx)
+                cov = tf.reduce_sum(cov, axis=idx + len(shape) - 1)
+            cov = tf.transpose(
+                cov, perm=perm_idxs + [i + len(perm_idxs) for i in perm_idxs]
+            )
+
+            if self.normalize:
+                norm = tf.reduce_sum(val)
+                jac = tf.eye(tf.shape(val)[0]) / norm - tf.expand_dims(val, 1) / norm**2
+                cov = tf.matmul(jac, tf.matmul(cov, tf.transpose(jac)))
+
+            var = tf.linalg.diag_part(cov)
```

Comment:
In case the data has a covariance matrix, in general the projected data will also have a covariance matrix with off diagonal terms.  This could be computed and propagated here.  (API would need to be extended I guess?)

Reply:
Yes, done